### ISSUE TITLE
Run commands with per-host command argument(s)

### ISSUE THREAD

**Issue by derchrisuk** [ParallelSSH/parallel-ssh#43]
Hi,

I'm currently looking for a way to run ssh sessions in parallel, and came across pssh.
But looking at the docs, it will only allow me to run the same command on many hosts.
Is it possible to run different commands on many hosts in parallel?

Thanks,
Christian

**Comment by pkittenis** [comment]
Hi there,

Thanks for the interest.

Currently it is not possible to run different commands on multiple hosts at the same time with one client - that is not a goal of this project.

Can use multiple clients though, ie:

```python
cmd1 = <..>
cmd2 = <..>
client1 = ParallelSSHClient(['myhost1', 'myhost2'])
client2 = ParallelSSHClient(['myhost3', 'myhost4'])
for client, cmd in [(client1, cmd1), (client2, cmd2)]:
    output = client.run_command(cmd)
    client.join(output)
```

Cmd1 is run on hosts `['myhost1', 'myhost2']` in parallel, cmd2 is run on `['myhost3', 'myhost4']` in parallel.

Running different commands in parallel on multiple hosts is a bit of an automation anti-pattern and is not likely to be supported.

Also, using multiple clients lets users programmatically decide when a set commands runs on which hosts which leads to consistent results.

**Comment by derchrisuk** [comment]
Thanks for the reply.

My scenario is that I would need to connect to several different VMs, and run a command with args.
But these args differ for each VMs

Currently I'm using just a plain multiprocessing, start a new Process with the required args, and append it to a jobs list for output buffering.

**Comment by pkittenis** [comment]
The internal function `_exec_command` can be used directly to do what you want in a single client - can see example usage of it at how [`run_command` calls it](https://github.com/pkittenis/parallel-ssh/blob/master/pssh/pssh_client.py#L288)

Multiple clients as above work just as well and do not use more than more thread/process.

As far as supporting that use case through a public interface - not likely to happen.

PRs are welcome of course if there is a sufficiently clear use case for it and does not detract from the stated goals of the project.

**Comment by pkittenis** [comment]
Note that this works out of the box for running multiple commands in parallel with one client:

```
for cmd in ['cmd1', 'cmd2']:
    client.run_command(cmd)
```

Output returned by `run_command` is for the hosts the cmd ran on and is therefore specific to both hosts and cmd.

**Comment by rokcarl** [comment]
I have a need to run different commands as well. Note that they're conceptually the same commands, they just have different params. E.g. on server-1 I'd like to run something like `run domain-1` and on server-2 I would run `run domain-2`.

What I want accomplished is the same, run the server. But the command is not the same.

**Comment by pkittenis** [comment]
Hi,

Can see the use case, sure. Have re-opened, think this can be provided and there seems to be value in it. 

Let me think how best to do this, suggestions welcome of course. Thinking that an ordered per-host argument list would suffice and still be optional.

Eg:

```
hosts = ['server-1', 'server-2']
client.run_command('run domain-%s', host_args=('1', '2',))
```

Host args will indicate the cmd does take string arguments and that the number of arguments supplied is correct. Library developer must ensure this is indeed the case.

**Comment by rokcarl** [comment]
This approach looks good. An alternative might be just having a `([host], [command])` tuple pair, e.g.`[('servers-1', 'restart domain1'), ('servers-2', 'restart domain2')]`.

**Comment by tedivm** [comment]
The need for this feature has come up multiple times for me but always in the same way- where the command I'm running needs an index identifier for each machine. For instance-

```shell
~/run_worker.py --worker=0
~/run_worker.py --worker=1
~/run_worker.py --worker=2
~/run_worker.py --worker=3
```

For me simply injecting the index of the host into the command would work great.

Alternatively, if this library had the option to add environmental variables into the per-host configuration this can be resolved while setting up the hosts (giving them their own identifier in the environmental variables) without the need to define arguments for each host and each command.

**Comment by pkittenis** [comment]
Great, sounds like the proposal will work for the above use case as well.

Regarding env variables - in general a library approach is preferred which allows for purely programmatic usage - a stated goal of this project - and also makes it possible for developers to use env variables to generate the library's arguments if they want.

The more general library approach leaves generation open to any implementation without having to develop explicit support for the varying implementations, of which there are many.

**Comment by tedivm** [comment]
Making this work with the idea I had above for "indexed" commands was remarkably easy.

```python
from pssh import ParallelSSHClient

class IndexedPSSH(ParallelSSHClient):

    def run_command(self, *args, **kwargs):
        stop_on_errors = kwargs.pop('stop_on_errors', True)
        cmds = []
        original_command = str(*args)
        for idx, host in enumerate(self.hosts):
            newcommand = original_command
            if '%s' in newcommand:
                newcommand = newcommand % (idx)
            cmds.append(self.pool.spawn(self._exec_command, host, newcommand, **kwargs))

        output = {}
        for cmd in cmds:
            try:
                self.get_output(cmd, output)
            except Exception:
                if stop_on_errors:
                    raise
        return output
```

The one change I would make to the code above (besides some basic cleanup perhaps) would be to fully replace all instances of `%s` with the index rather than just one.

**Comment by tedivm** [comment]
@pkittenis - any interest in a pull request that adds this functionality, similar to the extension I made above?

**Comment by pkittenis** [comment]
PR would be welcome, sure. It should be able to handle arbitrary number of args though to satisfy the other use cases. 

Like the idea about being able to use dict to format the string, sounds like that is worth using. Maybe host_args can be either tuple or dict, and used directly to format the cmd string, like with any other string.

**Comment by pkittenis** [comment]
Now in master, [see documentation](http://parallel-ssh.readthedocs.io/en/latest/pssh_client.html#pssh.pssh_client.ParallelSSHClient.run_command) for usage examples.

**Comment by nstudach** [comment]
Since the links are dead: Exmaple is now under:
https://parallel-ssh.readthedocs.io/en/latest/advanced.html?highlight=host.args#per-host-command-substitution
